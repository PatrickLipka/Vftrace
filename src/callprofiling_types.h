#ifndef CALLPROFILING_TYPES_H
#define CALLPROFILING_TYPES_H

typedef struct {
   // number of calls
   long long calls;
   // cycles spend in the function (including subfunctions)
   long long cycles;
   // time spend in the function (including subfunctions)
   long long time_usec;
   // time spend in the function (excluding subfunctions)
   // computed during final stack update
   long long time_excl_usec;
} callProfile_t;

#endif
