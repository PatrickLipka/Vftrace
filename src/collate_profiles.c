#include <stdlib.h>

#ifdef _MPI
#include <mpi.h>
#endif

#include "self_profile.h"
#include "stack_types.h"
#include "collated_stack_types.h"
#include "collate_callprofiles.h"
#ifdef _MPI
#include "collate_mpiprofiles.h"
#endif
#ifdef _OMP
#include "collate_ompprofiles.h"
#endif
#ifdef _CUPTI
#include "collate_cuptiprofiles.h"
#endif

void vftr_collate_profiles(collated_stacktree_t *collstacktree_ptr,
                           stacktree_t *stacktree_ptr) {
   SELF_PROFILE_START_FUNCTION;
   int myrank;
   int nranks;
   int nprofiles = stacktree_ptr->nstacks;
   int *nremote_profiles = NULL;
#ifdef _MPI
   PMPI_Comm_rank(MPI_COMM_WORLD, &myrank);
   PMPI_Comm_size(MPI_COMM_WORLD, &nranks);
#else
   myrank = 0;
   nranks = 1;
#endif

   // get the number of profiles to collect from each rank
   if (myrank == 0) {
      nremote_profiles = (int*) malloc(nranks*sizeof(int));
   }
#ifdef _MPI
   PMPI_Gather(&nprofiles, 1,
               MPI_INT,
               nremote_profiles, 1,
               MPI_INT,
               0, MPI_COMM_WORLD);
#else
   nremote_profiles[0] = nprofiles;
#endif

   vftr_collate_callprofiles(collstacktree_ptr, stacktree_ptr,
                             myrank, nranks, nremote_profiles);
#ifdef _MPI
   vftr_collate_mpiprofiles(collstacktree_ptr, stacktree_ptr,
                            myrank, nranks, nremote_profiles);
#endif

#ifdef _OMP
   vftr_collate_ompprofiles(collstacktree_ptr, stacktree_ptr,
                            myrank, nranks, nremote_profiles);
#endif

#ifdef _CUPTI
   vftr_collate_cuptiprofiles(collstacktree_ptr, stacktree_ptr,
                              myrank, nranks, nremote_profiles);
#endif

   free(nremote_profiles);
   SELF_PROFILE_END_FUNCTION;
}
