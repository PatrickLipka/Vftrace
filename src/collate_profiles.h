#ifndef COLLATE_PROFILES_H
#define COLLATE_PROFILES_H

#include <stdlib.h>

#include "stack_types.h"
#include "collated_stack_types.h"

void vftr_collate_profiles(collated_stacktree_t *collstacktree_ptr,
                           stacktree_t *stacktree_ptr);

#endif
