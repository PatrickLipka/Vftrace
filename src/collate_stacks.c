#include <stdlib.h>

#include <string.h>
#ifdef _MPI
#include <mpi.h>
#endif

#include "stack_types.h"
#include "collated_hash_types.h"
#include "collated_stack_types.h"

#include "hashing.h"
#include "collate_hashes.h"
#include "search.h"

collated_stacktree_t vftr_new_collated_stacktree() {
   collated_stacktree_t stacktree;
   stacktree.nstacks = 0;
   stacktree.stacks = NULL;
   return stacktree;
}

#ifdef _MPI
void vftr_broadcast_collated_stacktree_root(collated_stacktree_t *stacktree_ptr) {
   int nranks;
   PMPI_Comm_size(MPI_COMM_WORLD, &nranks);
   int nstacks = stacktree_ptr->nstacks;
   // broadcasting the caller ids
   int *tmpintarr = (int*) malloc((nstacks+1)*sizeof(int));
   for (int istack=0; istack<nstacks; istack++) {
      tmpintarr[istack] = stacktree_ptr->stacks[istack].caller;
   }
   PMPI_Bcast(tmpintarr, nstacks, MPI_INT, 0, MPI_COMM_WORLD);
   // communicate the precise-tracing status for each function
   for (int istack=0; istack<nstacks; istack++) {
      tmpintarr[istack] = stacktree_ptr->stacks[istack].precise ? 1 : 0;
   }
   PMPI_Bcast(tmpintarr, nstacks, MPI_INT, 0, MPI_COMM_WORLD);
   // broadcasting the function names.
   // first the length of each name
   int totallen = 0;
   for (int istack=0; istack<nstacks; istack++) {
      tmpintarr[istack] = strlen(stacktree_ptr->stacks[istack].name);
      totallen += tmpintarr[istack];
      totallen += 1; // null terminator between all strings
   }
   // the total length is appended to the tmpintarr
   tmpintarr[nstacks] = totallen;
   PMPI_Bcast(tmpintarr, nstacks+1, MPI_INT, 0, MPI_COMM_WORLD);
   char *tmpchararr = (char*) malloc(totallen*sizeof(char));
   char *charptr = tmpchararr;
   for (int istack=0; istack<nstacks; istack++) {
      strcpy(charptr, stacktree_ptr->stacks[istack].name);
      charptr += tmpintarr[istack];
      charptr++; // null terminator should be copied by strcpy
   }
   PMPI_Bcast(tmpchararr, totallen, MPI_CHAR, 0, MPI_COMM_WORLD);
   free(tmpintarr);
   free(tmpchararr);
}

void vftr_broadcast_collated_stacktree_receivers(collated_stacktree_t *stacktree_ptr) {
   int nranks;
   PMPI_Comm_size(MPI_COMM_WORLD, &nranks);
   int nstacks = stacktree_ptr->nstacks;
   // receiving the caller ids
   int *tmpintarr = (int*) malloc((nstacks+1)*sizeof(int));
   PMPI_Bcast(tmpintarr, nstacks, MPI_INT, 0, MPI_COMM_WORLD);
   for (int istack=0; istack<nstacks; istack++) {
      stacktree_ptr->stacks[istack].caller = tmpintarr[istack];
   }
   // recive the precise-tracing status for each function
   PMPI_Bcast(tmpintarr, nstacks, MPI_INT, 0, MPI_COMM_WORLD);
   for (int istack=0; istack<nstacks; istack++) {
      stacktree_ptr->stacks[istack].precise = tmpintarr[istack] == 1 ? true : false;
   }
   // receiving the function names;
   // first the length of each name;
   int totallen = 0;
   PMPI_Bcast(tmpintarr, nstacks+1, MPI_INT, 0, MPI_COMM_WORLD);
   // the total length is appended to the tmpintarr
   totallen = tmpintarr[nstacks];
   char *tmpchararr = (char*) malloc(totallen*sizeof(char));
   char *charptr = tmpchararr;
   PMPI_Bcast(tmpchararr, totallen, MPI_CHAR, 0, MPI_COMM_WORLD);
   for (int istack=0; istack<nstacks; istack++) {
      stacktree_ptr->stacks[istack].name = strdup(charptr);
      charptr += tmpintarr[istack];
      charptr++; // null terminator should be copied by strcpy
   }
   free(tmpintarr);
   free(tmpchararr);
}

void vftr_broadcast_collated_stacktree(collated_stacktree_t *stacktree_ptr) {
   int nranks;
   PMPI_Comm_size(MPI_COMM_WORLD, &nranks);
   if (nranks > 1) {
      int myrank;
      PMPI_Comm_rank(MPI_COMM_WORLD, &myrank);
      if (myrank == 0) {
         vftr_broadcast_collated_stacktree_root(stacktree_ptr);
      } else {
         vftr_broadcast_collated_stacktree_receivers(stacktree_ptr);
      }
   }
}
#endif

collated_stacktree_t vftr_collate_stacks(stacktree_t *stacktree_ptr) {
   collated_stacktree_t coll_stacktree;

   // first compute the hashes for all stacks
   vftr_compute_stack_hashes(stacktree_ptr);

   // collate hashes between processes
   hashlist_t hashlist = vftr_collate_hashes(stacktree_ptr);

   // create empty collated stacktree
   coll_stacktree.nstacks = hashlist.nhashes;
   coll_stacktree.stacks = (collated_stack_t*)
      malloc(coll_stacktree.nstacks*sizeof(collated_stack_t));
   for (int istack=0; istack<coll_stacktree.nstacks; istack++) {
      coll_stacktree.stacks[istack].local_stack = NULL;
      coll_stacktree.stacks[istack].gid = istack;
      coll_stacktree.stacks[istack].precise = false;
      coll_stacktree.stacks[istack].caller = -1;
      coll_stacktree.stacks[istack].name = NULL;
      coll_stacktree.stacks[istack].hash = hashlist.hashes[istack];
   }

   // build a lookup table each to translate local2global and global2local
   int *local2global_ID = (int*) malloc(stacktree_ptr->nstacks*sizeof(int));
   int *global2local_ID = (int*) malloc(coll_stacktree.nstacks*sizeof(int));
   for (int istack=0; istack<stacktree_ptr->nstacks; istack++) {
      // -1 in the lookup table means that the local stack does not exist
      local2global_ID[istack] = -1;
   }
   for (int istack=0; istack<coll_stacktree.nstacks; istack++) {
      // -1 in the lookup table means that the local stack does not exist
      global2local_ID[istack] = -1;
   }
   // assign every function its global ID
   // the global ID is the index in the hash table
   // fill the looup tables
   for (int istack=0; istack<stacktree_ptr->nstacks; istack++) {
      int idx = vftr_binary_search_uint64(hashlist.nhashes,
                                          hashlist.hashes,
                                          stacktree_ptr->stacks[istack].hash);
      if (idx >= 0) {
         stacktree_ptr->stacks[istack].gid = idx;
         local2global_ID[istack] = idx;
         global2local_ID[idx] = istack;
      }
   }

   // hashtable is no longer needed
   vftr_collated_hashlist_free(&hashlist);

   // fill in the locally known information
   // special treatment for init because it does not have a caller
   int myrank = 0;
#ifdef _MPI
   int mpi_initialized;
   PMPI_Initialized(&mpi_initialized);
   if (mpi_initialized) {
      PMPI_Comm_rank(MPI_COMM_WORLD, &myrank);
   }
#endif
   if (myrank == 0) {
      {
         int lid = 0;
         stack_t *local_stack = stacktree_ptr->stacks+lid;
         int gid = local2global_ID[lid];
         collated_stack_t *global_stack = coll_stacktree.stacks+gid;
         global_stack->local_stack = local_stack;
         global_stack->gid = gid;
         global_stack->precise = local_stack->precise;
         global_stack->caller = -1;
         global_stack->name = strdup(local_stack->name);
      }
      for (int lid=1; lid<stacktree_ptr->nstacks; lid++) {
         stack_t *local_stack = stacktree_ptr->stacks+lid;
         int gid = local2global_ID[lid];
         collated_stack_t *global_stack = coll_stacktree.stacks+gid;
         global_stack->local_stack = local_stack;
         global_stack->gid = gid;
         global_stack->precise = local_stack->precise;
         global_stack->caller = local2global_ID[local_stack->caller];
         global_stack->name = strdup(local_stack->name);
      }
   }

#ifdef _MPI
   if (mpi_initialized) {
      if (myrank == 0) {
         // if there are multiple processes the table might still be missing entries
         int *missingstacks = (int*) malloc(coll_stacktree.nstacks*sizeof(int));
         // loop over all ranks and collect the missing information
         int nranks;
         PMPI_Comm_size(MPI_COMM_WORLD, &nranks);
         for (int irank=1; irank<nranks; irank++) {
            int nmissing = 0;
            for (int istack=0; istack<coll_stacktree.nstacks; istack++) {
               if (coll_stacktree.stacks[istack].name == NULL) {
                  missingstacks[nmissing] = istack;
                  nmissing++;
               }
            }

            // Send to the selected process how many entries are still missing
            PMPI_Send(&nmissing, 1, MPI_INT, irank, 0, MPI_COMM_WORLD);
            // if at least one entry is missing proceed
            if (nmissing > 0) {
               // Send the missing IDs
               PMPI_Send(missingstacks, nmissing, MPI_INT, irank, 0, MPI_COMM_WORLD);
               // Receive how many missing stacks the other process can fill in
               MPI_Status mystat;
               int hasnmissing;
               PMPI_Recv(&hasnmissing, 1, MPI_INT, irank, 0, MPI_COMM_WORLD, &mystat);
               // only proceed if the number of stacks is positive
               if (hasnmissing > 0) {
                   // Allocate space for the Stack information
                   // globalIDs  -> 4*istack+0
                   // returnID   -> 4*istack+1
                   // NameLength -> 4*istack+2
                   // precise    -> 4*istack+3
                   int *missingStackInfo = (int*) malloc(4*hasnmissing*sizeof(int));
                   // Receive the found information from remote process
                   PMPI_Recv(missingStackInfo, 4*hasnmissing, MPI_INT,
                             irank, 0, MPI_COMM_WORLD, &mystat);

                   // Create a buffer that contains all stack names in contatenated form
                   int sumlength = 0;
                   for (int istack=0; istack<hasnmissing; istack++) {
                      sumlength += missingStackInfo[4*istack+2];
                   }
                   char *concatNames = (char*) malloc(sumlength*sizeof(char));

                   // Receive the concatenated String
                   PMPI_Recv(concatNames, sumlength, MPI_CHAR,
                             irank, 0, MPI_COMM_WORLD, &mystat);

                   // Write all the gathered info to the global stackinfo
                   char *tmpstrptr = concatNames;
                   for (int istack=0; istack<hasnmissing; istack++) {
                      int globID = missingStackInfo[4*istack+0];
                      coll_stacktree.stacks[globID].caller = missingStackInfo[4*istack+1];
                      coll_stacktree.stacks[globID].name = strdup(tmpstrptr);
                      // next string
                      tmpstrptr += missingStackInfo[4*istack+2];
                      // is precise ?
                      coll_stacktree.stacks[globID].precise = missingStackInfo[4*istack+3] != 0;
                   }

                   free(concatNames);
                   free(missingStackInfo);

               }
            }
         }
         free(missingstacks);
      } else {
         // not process 0
         MPI_Status mystat;
         // receive how many entries process 0 is missing
         int nmissing;
         PMPI_Recv(&nmissing, 1, MPI_INT, 0, 0, MPI_COMM_WORLD, &mystat);
         // if at least one entry is missing proceed
         if (nmissing > 0) {
            // allocate space to hold the missing ids
            int *missingstacks = (int*) malloc(nmissing*sizeof(int));
            // receiving the missing stacks
            PMPI_Recv(missingstacks, nmissing, MPI_INT, 0, 0, MPI_COMM_WORLD, &mystat);

            // check how many of the missing stacks this process has infos about
            int hasnmissing = 0;
            for (int imissing=0; imissing<nmissing; imissing++) {
               int globID = missingstacks[imissing];
               if (global2local_ID[globID] >= 0) {
                  hasnmissing++;
               }
            }
            // Report back how many missing stacks this process can fill in
            PMPI_Send(&hasnmissing, 1, MPI_INT, 0, 0, MPI_COMM_WORLD);
            // only proceed if the number of stacks is positive
            if (hasnmissing > 0) {
               // Allocate space for the global IDs, return ID, and the name lengths
               // globalIDs  -> 4*istack+0
               // returnID   -> 4*istack+1
               // NameLength -> 4*istack+2
               // Precise    -> 4*istack+3
               int *missingStackInfo = (int*) malloc(4*hasnmissing*sizeof(int));
               // Go through the stacks and record the needed information
               int imatch = 0;
               for (int istack=0; istack<nmissing; istack++) {
                  int globID = missingstacks[istack];
                  int locID = global2local_ID[globID];
                  if (locID >= 0) {
                     missingStackInfo[4*imatch+0] = globID;
                     stack_t *local_stack = stacktree_ptr->stacks+locID;
                     missingStackInfo[4*imatch+1] = local2global_ID[local_stack->caller];
                     // add one to length due to null terminator
                     missingStackInfo[4*imatch+2] = strlen(local_stack->name) + 1;
                     missingStackInfo[4*imatch+3] = local_stack->precise ? 1 : 0;
                     imatch++;
                  }
               }
               // Communicate the found information to process 0;
               PMPI_Send(missingStackInfo, 4*hasnmissing, MPI_INT, 0, 0, MPI_COMM_WORLD);
               // Create a buffer that contains all stack names in contatenated form
               int sumlength = 0;
               for (int istack=0; istack<hasnmissing; istack++) {
                  sumlength += missingStackInfo[4*istack+2];
               }
               char *concatNames = (char*) malloc(sumlength*sizeof(char));
               // concatenate the names into one string
               char *tmpstrptr = concatNames;
               for (int istack=0; istack<hasnmissing; istack++) {
                  int globID = missingStackInfo[4*istack+0];
                  int locID = global2local_ID[globID];
                  strcpy(tmpstrptr, stacktree_ptr->stacks[locID].name);
                  tmpstrptr += missingStackInfo[4*istack+2]-1;
                  // add null terminator
                  *tmpstrptr = '\0';
                  tmpstrptr++;
               }
               // communicate the concatenated string to process 0
               PMPI_Send(concatNames, sumlength, MPI_CHAR, 0, 0, MPI_COMM_WORLD);

               // free everything. This should be all on the remote processes
               free(concatNames);
               free(missingStackInfo);
            }
            free(missingstacks);
         }
      }

      vftr_broadcast_collated_stacktree(&coll_stacktree);
   }
#endif

   free(local2global_ID);
   free(global2local_ID);

   return coll_stacktree;
}

void vftr_collated_stacktree_free(collated_stacktree_t *stacktree_ptr) {
   if (stacktree_ptr->nstacks > 0) {
      for (int istack=0; istack<stacktree_ptr->nstacks; istack++) {
         free(stacktree_ptr->stacks[istack].name);
      }
      free(stacktree_ptr->stacks);
      stacktree_ptr->stacks = NULL;
      stacktree_ptr->nstacks = 0;
   }
}

char *vftr_get_collated_stack_string(collated_stacktree_t stacktree, int stackid) {
   int stringlen = 0;
   int tmpstackid = stackid;
   stringlen += strlen(stacktree.stacks[stackid].name);
   stringlen ++; // function seperating character "<", or null terminator
//   if (stacktree.stacks[tmpstackid].precise) {
//      stringlen ++; // '*' for indicating precise functions
//   }
   while (stacktree.stacks[tmpstackid].caller >= 0) {
      tmpstackid = stacktree.stacks[tmpstackid].caller;
      stringlen += strlen(stacktree.stacks[tmpstackid].name);
      stringlen ++; // function seperating character "<", or null terminator
//      if (stacktree.stacks[tmpstackid].precise) {
//         stringlen ++; // '*' for indicating precise functions
//      }
   }
   char *stackstring = (char*) malloc(stringlen*sizeof(char));
   // copy the chars one by one so there is no need to call strlen again.
   // thus minimizing reading the same memory locations over and over again.
   tmpstackid = stackid;
   char *tmpname_ptr = stacktree.stacks[tmpstackid].name;
   char *tmpstackstring_ptr = stackstring;
   while (*tmpname_ptr != '\0') {
      *tmpstackstring_ptr = *tmpname_ptr;
      tmpstackstring_ptr++;
      tmpname_ptr++;
   }
//   if (stacktree.stacks[tmpstackid].precise) {
//      *tmpstackstring_ptr = '*';
//      tmpstackstring_ptr++;
//   }
   while (stacktree.stacks[tmpstackid].caller >= 0) {
      // add function name separating character
      *tmpstackstring_ptr = '<';
      tmpstackstring_ptr++;
      tmpstackid = stacktree.stacks[tmpstackid].caller;
      char *tmpname_ptr = stacktree.stacks[tmpstackid].name;
      while (*tmpname_ptr != '\0') {
         *tmpstackstring_ptr = *tmpname_ptr;
         tmpstackstring_ptr++;
         tmpname_ptr++;
      }
//      if (stacktree.stacks[tmpstackid].precise) {
//         *tmpstackstring_ptr = '*';
//         tmpstackstring_ptr++;
//      }
   }
   // replace last char with a null terminator
   *tmpstackstring_ptr = '\0';
   return stackstring;
}

void vftr_print_collated_stack(FILE *fp, collated_stacktree_t stacktree, int stackid) {
   char *stackstr = vftr_get_collated_stack_string(stacktree, stackid);
   fprintf(fp, "%s", stackstr);
   free(stackstr);
}

void vftr_print_collated_stacklist(FILE *fp, collated_stacktree_t stacktree) {
   for (int istack=0; istack<stacktree.nstacks; istack++) {
      fprintf(fp, "%u: ", istack);
      vftr_print_collated_stack(fp, stacktree, istack);
      fprintf(fp, "\n");
   }
}
