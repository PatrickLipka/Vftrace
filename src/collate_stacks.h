#ifndef COLLATE_STACKS_H
#define COLLATE_STACKS_H

#include "stack_types.h"
#include "collated_stack_types.h"

collated_stacktree_t vftr_new_collated_stacktree();

collated_stacktree_t vftr_collate_stacks(stacktree_t *stacktree_ptr);

void vftr_collated_stacktree_free(collated_stacktree_t *stacktree_ptr);

void vftr_print_collated_stacklist(FILE *fp, collated_stacktree_t stacktree);

char *vftr_get_collated_stack_string(collated_stacktree_t stacktree, int stackid);

#endif
