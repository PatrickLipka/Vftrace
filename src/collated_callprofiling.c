#include <stdlib.h>
#include <stdio.h>

#include "self_profile.h"
#include "collated_callprofiling_types.h"
#include "collated_stack_types.h"
#include "stack_types.h"

#ifdef _MPI
#include <mpi.h>
#endif

collated_callProfile_t vftr_new_collated_callprofiling() {
   SELF_PROFILE_START_FUNCTION;
   collated_callProfile_t prof;
   prof.calls = 0ll;
   prof.time_usec = 0ll;
   prof.time_excl_usec = 0ll;
   prof.overhead_usec = 0ll;
   prof.on_nranks = 0;
   prof.max_on_rank = 0;
   prof.min_on_rank = 0;
   prof.average_time_usec = 0ll;
   prof.max_time_usec = 0ll;
   prof.min_time_usec = 0ll;
   prof.max_imbalance = 0.0;
   prof.max_imbalance_on_rank = 0;
   SELF_PROFILE_END_FUNCTION;
   return prof;
}

long long vftr_get_total_collated_call_overhead(collated_stacktree_t stacktree) {
   SELF_PROFILE_START_FUNCTION;
   // accumulate the hook overhead for each thread separately
   long long overheads_usec = 0ll;

   int nstacks = stacktree.nstacks;
   for (int istack=0; istack<nstacks; istack++) {
      collated_stack_t *stack = stacktree.stacks+istack;
      collated_profile_t *prof = &(stack->profile);
      overheads_usec += prof->callProf.overhead_usec;
   }
   SELF_PROFILE_END_FUNCTION;
   return overheads_usec;
}

void vftr_collated_callprofiling_free(collated_callProfile_t *callprof_ptr) {
   SELF_PROFILE_START_FUNCTION;
   (void) callprof_ptr;
   SELF_PROFILE_END_FUNCTION;
}

void vftr_print_collated_callprofiling(FILE *fp, collated_callProfile_t callprof) {
   fprintf(fp, "calls: %lld, time(incl/excl): %lld/%lld (overhead: %lld)\n",
           callprof.calls, callprof.time_usec, callprof.time_excl_usec,
           callprof.overhead_usec);
}
