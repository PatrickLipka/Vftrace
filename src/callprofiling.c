#include <stdlib.h>
#include <stdio.h>

#include "self_profile.h"
#include "callprofiling_types.h"
#include "collated_stack_types.h"
#include "stack_types.h"

#ifdef _MPI
#include <mpi.h>
#endif

callProfile_t vftr_new_callprofiling() {
   SELF_PROFILE_START_FUNCTION;
   callProfile_t prof;
   prof.calls = 0ll;
   prof.time_usec = 0ll;
   prof.time_excl_usec = 0ll;
   prof.overhead_usec = 0ll;
   SELF_PROFILE_END_FUNCTION;
   return prof;
}

void vftr_accumulate_callprofiling(callProfile_t *prof,
                                   int calls,
                                   long long time_usec) {
   prof->calls += calls;
   prof->time_usec += time_usec;
}

void vftr_accumulate_callprofiling_overhead(callProfile_t *prof,
                                            long long overhead_usec) {
   prof->overhead_usec += overhead_usec;
}

void vftr_update_stacks_exclusive_time(stacktree_t *stacktree_ptr) {
   SELF_PROFILE_START_FUNCTION;
   int nstacks = stacktree_ptr->nstacks;
   stack_t *stacks = stacktree_ptr->stacks;
   // exclusive time for init is 0, therefore it does not need to be computed.
   for (int istack=1; istack<nstacks; istack++) {
      stack_t *mystack = stacks + istack;
      // need to go over the calling profiles threadwise
      for (int iprof=0; iprof<mystack->profiling.nprofiles; iprof++) {
         profile_t *myprof = mystack->profiling.profiles+iprof;
         myprof->callProf.time_excl_usec = myprof->callProf.time_usec;
         // subtract the time spent in the callees
         for (int icallee=0; icallee<mystack->ncallees; icallee++) {
            int calleeID = mystack->callees[icallee];
            stack_t *calleestack = stacks+calleeID;
            // search for a thread matching profile in the callee profiles
            int calleprofID = -1;
            for (int jprof=0; jprof<calleestack->profiling.nprofiles; jprof++) {
               profile_t *calleeprof = calleestack->profiling.profiles+jprof;
               if (myprof->threadID == calleeprof->threadID) {
                  calleprofID = jprof;
                  break;
               }
            }
            // a matching callee profile was found
            if (calleprofID >= 0) {
               profile_t *calleeprof = calleestack->profiling.profiles+calleprofID;
               myprof->callProf.time_excl_usec -= calleeprof->callProf.time_usec;
            }
         }
      }
   }
   SELF_PROFILE_END_FUNCTION;
}

long long *vftr_get_total_call_overhead(stacktree_t stacktree, int nthreads) {
   SELF_PROFILE_START_FUNCTION;
   // accumulate the hook overhead for each thread separately
   long long *overheads_usec = (long long*) malloc(nthreads*sizeof(long long));
   for (int ithread=0; ithread<nthreads; ithread++) {
      overheads_usec[ithread] = 0ll;
   }

   int nstacks = stacktree.nstacks;
   for (int istack=0; istack<nstacks; istack++) {
      stack_t *stack = stacktree.stacks+istack;
      int nprofs = stack->profiling.nprofiles;
      for (int iprof=0; iprof<nprofs; iprof++) {
         profile_t *prof = stack->profiling.profiles+iprof;
         int threadID = prof->threadID;
         overheads_usec[threadID] += prof->callProf.overhead_usec;
      }
   }
   SELF_PROFILE_END_FUNCTION;
   return overheads_usec;
}

long long vftr_get_total_collated_call_overhead(collated_stacktree_t stacktree) {
   SELF_PROFILE_START_FUNCTION;
   // accumulate the hook overhead for each thread separately
   long long overheads_usec = 0ll;

   int nstacks = stacktree.nstacks;
   for (int istack=0; istack<nstacks; istack++) {
      collated_stack_t *stack = stacktree.stacks+istack;
      collated_profile_t *prof = &(stack->profile);
      overheads_usec += prof->callProf.overhead_usec;
   }
   SELF_PROFILE_END_FUNCTION;
   return overheads_usec;
}

void vftr_callprofiling_free(callProfile_t *callprof_ptr) {
   SELF_PROFILE_START_FUNCTION;
   (void) callprof_ptr;
   SELF_PROFILE_END_FUNCTION;
}

void vftr_print_callprofiling(FILE *fp, callProfile_t callprof) {
   fprintf(fp, "calls: %lld, time(incl/excl): %lld/%lld (overhead: %lld)\n",
           callprof.calls, callprof.time_usec, callprof.time_excl_usec,
           callprof.overhead_usec);
}
