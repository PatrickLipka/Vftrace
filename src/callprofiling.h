#ifndef CALLPROFILING_H
#define CALLPROFILING_H

#include <stdio.h>

#include "callprofiling_types.h"
#include "stack_types.h"

#include "callprofiling.h"

callProfile_t vftr_new_callprofiling();

void vftr_accumulate_callprofiling(callProfile_t *prof,
                                   int calls,
                                   long long cycles,
                                   long long time_usec);

void vftr_callprofiling_free(callProfile_t *callprof_ptr);

void vftr_update_stacks_exclusive_time(int nstacks, stack_t *stacks);

void vftr_print_callprofiling(FILE *fp, callProfile_t callprof);

#endif
