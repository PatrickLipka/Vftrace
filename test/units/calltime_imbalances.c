#include <stdlib.h>
#include <stdio.h>

#include "self_profile.h"
#include "environment_types.h"
#include "environment.h"
#include "symbol_types.h"
#include "symbols.h"
#include "stack_types.h"
#include "stacks.h"
#include "profiling_types.h"
#include "profiling.h"
#include "callprofiling_types.h"
#include "callprofiling.h"
#include "collated_callprofiling_types.h"
#include "collated_callprofiling.h"
#include "collated_stack_types.h"
#include "collate_stacks.h"
#include "collate_profiles.h"

#include "dummy_stacktree.h"

#ifdef _MPI
#include <mpi.h>
#endif

int main(int argc, char **argv) {
   INIT_SELF_PROF_VFTRACE;
#if defined(_MPI)
   PMPI_Init(&argc, &argv);
#else
   (void) argc;
   (void) argv;
#endif

   stacktree_t stacktree = vftr_init_dummy_stacktree (20, 0);

   vftr_register_dummy_stack(&stacktree, "func0<init", 0, 10, 2);
   vftr_register_dummy_stack(&stacktree, "func0<init", 1, 10, 4);
   vftr_register_dummy_stack(&stacktree, "func0<init", 2, 10, 8);
   vftr_register_dummy_stack(&stacktree, "func0<init", 3, 10, 16);
   vftr_register_dummy_stack(&stacktree, "func0<init", 4, 10, 32);
   vftr_register_dummy_stack(&stacktree, "func0<init", 5, 10, 64);

   vftr_register_dummy_stack(&stacktree, "func1<init", 0, 10, 128);
   vftr_register_dummy_stack(&stacktree, "func1<init", 1, 10, 256);
   vftr_register_dummy_stack(&stacktree, "func1<init", 2, 10, 515);

   vftr_register_dummy_stack(&stacktree, "func2<func1<init", 1, 3, 1024);
   vftr_register_dummy_stack(&stacktree, "func2<func1<init", 2, 4, 2048);

   vftr_register_dummy_stack(&stacktree, "func3<func0<init", 0, 3, 4096);
   vftr_register_dummy_stack(&stacktree, "func3<func0<init", 2, 2, 8192);
   vftr_register_dummy_stack(&stacktree, "func3<func0<init", 4, 3, 16384);

   vftr_register_dummy_stack(&stacktree, "func4<func0<init", 1, 4, 32768);
   vftr_register_dummy_stack(&stacktree, "func4<func0<init", 3, 5, 65536);
   vftr_register_dummy_stack(&stacktree, "func4<func0<init", 5, 6, 131072);

   vftr_register_dummy_stack(&stacktree, "func5<func4<func0<init", 5, 2, 262144);

   vftr_update_stacks_exclusive_time(&stacktree);

   // collate stacks to get the global ID
   collated_stacktree_t collated_stacktree = vftr_collate_stacks(&stacktree);
   vftr_collate_profiles(&collated_stacktree, &stacktree);

   for (int istack=0; istack<collated_stacktree.nstacks; istack++) {
      char *stackstr = vftr_get_collated_stack_string(collated_stacktree, istack, false);
      fprintf(stdout, "%s:\n   ", stackstr);
      vftr_print_calltime_imbalances(stdout,
         collated_stacktree.stacks[istack].profile.callprof);
      free(stackstr);
   }

   vftr_stacktree_free(&stacktree);
   vftr_collated_stacktree_free(&collated_stacktree);

#ifdef _MPI
   PMPI_Finalize();
#endif

   FINALIZE_SELF_PROF_VFTRACE;
   return 0;
}
