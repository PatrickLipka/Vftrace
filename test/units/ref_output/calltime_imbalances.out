func2<func1<init:
   avg: 7us, min/max=7us(on 0)/7us(on 0), imb=  0.00% on 0
func1<init:
   avg: 23us, min/max=23us(on 0)/23us(on 0), imb=  0.00% on 0
func3<func0<init:
   avg: 8us, min/max=8us(on 0)/8us(on 0), imb=  0.00% on 0
init:
   avg: 0us, min/max=0us(on 0)/0us(on 0), imb=  0.00% on 0
func0<init:
   avg: 37us, min/max=37us(on 0)/37us(on 0), imb=  0.00% on 0
func4<func0<init:
   avg: 13us, min/max=13us(on 0)/13us(on 0), imb=  0.00% on 0
func5<func4<func0<init:
   avg: 2us, min/max=2us(on 0)/2us(on 0), imb=  0.00% on 0
